classes/various/extInteger.sc

```diff
@@ -0,0 +1,158 @@
+/*
+	You should have received a copy of the GNU General Public License along with the
+	SuperCollider3 version of the Ambisonic Toolkit (ATK). If not, see
+	<http://www.gnu.org/licenses/>.
+*/
+
+//---------------------------------------------------------------------
+//	MathLib library
+//	https://github.com/supercollider-quarks/MathLib
+//
+// 	Extension: Integer
+//---------------------------------------------------------------------
+
+
++ Integer {
+
+	/*
+	Perfect Power algorithms adapted from those found in:
+
+	Title: Detecting perfect powers in essentially linear time
+	Author: Daniel J. Bernstein
+	Journal: Math. Comp. 67 (1998), 1253-1283
+
+	https://www.ams.org/journals/mcom/1998-67-223/S0025-5718-98-00952-1/home.html
+
+	See also:
+
+	Title: Primality Testing - Theory, Complexity, and Applications
+	Author: Riley Worthington
+
+	https://www.whitman.edu/documents/Academics/Mathematics/2018/Worthington.pdf
+
+
+	NOTE:
+	It is expected the below algorithms may be further optimized by refactoring
+	in terms of Bernstein's 2-adic variant primitives (see: PART VI. PRACTICAL
+	IMPROVEMENTS). Doing so requires implementation of b-bit approximate kth
+	power of r (see: 6. Approximate powers) and related truncated arithmetic.
+	*/
+
+	// perfect power: detection - unknown base, exponent
+	isPerfectPower {
```

Comment:
Better use the `and`instead of the `&&` and bracket the second operand in a function. This will make the call more efficient.


```supercollider
// compare
{  a > 2  and: a.isPowerOfTwo }.def.dumpByteCodes // always calls isPowerOfTwo
{  a > 2  && { a.isPowerOfTwo } }.def.dumpByteCodes // PushLiteralX instance of FunctionDef - closed
{  a > 2  and: { a.isPowerOfTwo } }.def.dumpByteCodes // JumpIfFalsePushFalse 4
```

But you don't need it here, just write something like this:

```
isPerfectPower {
	^if(this < 2) { this.isPowerOfTwo } { this.perfectPower.notNil }
}
```

with regards to the line `this.perfectPower.notNil` see below.

Reply:
Ok, for this one, I've re-written as:

```
isPerfectPower {
	^if( this.isPowerOfTwo and: { this > 3 }, {
		true
	}, {
		this.perfectPower.notNil
	})
}
```

The notion is that given `Integer:-isPowerOfTwo` is implemented via a primitive, this _should_ be quick and efficient. But as the [usual convention](https://en.wikipedia.org/wiki/Perfect_power) for perfect powers is to start with 4, we need to check for that. I've expressed this as > 3, but >=4 could also be used.

Otherwise, try to return a perfect power.

Reply:
good! Just out of curiosity: why do you find this better than the one-liner I suggested above? To me it reads better, but my view may be skewed ...

Reply:
Doh! (Long week!)

Ok, here's a one liner version that does the job:

```
isPerfectPower {
	^if(this.isPowerOfTwo) { this > 3 } { this.perfectPower.notNil }
}
```

Thanks for the suggestion, it does read clearer. I'll go ahead and make this change.

Reply:
ah, now I understand, this is because of the convention that perfect powers is to start with 4 (I didn't know this, sorry!). In the case that it is `<= 3`, we shouldn't have to call `perfectPower` at all then.

```supercollider
isPerfectPower {
	^(this > 3) and: { this.isPowerOfTwo or: { this.perfectPower.notNil } }
}

``` 

but maybe it reads better to write:

```supercollider
isPerfectPower {
	if(x < 4) { ^false };
	^this.isPowerOfTwo or: { this.perfectPower.notNil }
}
```

(I'm sorry, I hope this isn't annoying! Just trying to perfect it …)

Reply:
No worries... this week has been fairly "brain fogged" for me. (I blame COVID.)

Yes, this form is much better and more expressive. (I'm ashamed to say that Logic Design was one of my favorite classes as an undergrad.)

Reply:
it is fun, indeed! :)